File: typescript/src/bridge/bridge.test.ts
import { describe, expect, it } from "vitest";
import { Abstraction } from "./abstraction.ts";
import { ConcreteImplementorA } from "./concrete-implementor-a.ts";
import { ConcreteImplementorB } from "./concrete-implementor-b.ts";
import { RefinedAbstraction } from "./refined-abstraction.ts";

describe("Bridge", () => {
  it("any abstraction works with any implementor", () => {
    expect(new Abstraction(new ConcreteImplementorA()).operation()).toBe("Abstraction(ConcreteImplementorA)");
    expect(new Abstraction(new ConcreteImplementorB()).operation()).toBe("Abstraction(ConcreteImplementorB)");
    expect(new RefinedAbstraction(new ConcreteImplementorA()).operation()).toBe("RefinedAbstraction(ConcreteImplementorA)");
    expect(new RefinedAbstraction(new ConcreteImplementorB()).operation()).toBe("RefinedAbstraction(ConcreteImplementorB)");
  });

  it("a new implementor needs no change on the abstraction side", () => {
    const custom = { operationImpl: () => "Custom" };
    expect(new Abstraction(custom).operation()).toBe("Abstraction(Custom)");
    expect(new RefinedAbstraction(custom).operation()).toBe("RefinedAbstraction(Custom)");
  });
});

File: java/src/test/java/com/penapereira/patterns/bridge/BridgeTest.java
package com.penapereira.patterns.bridge;

import static org.junit.jupiter.api.Assertions.assertEquals;

import org.junit.jupiter.api.Test;

class BridgeTest {

    @Test
    void anyAbstractionWorksWithAnyImplementor() {
        assertEquals("Abstraction(ConcreteImplementorA)", new Abstraction(new ConcreteImplementorA()).operation());
        assertEquals("Abstraction(ConcreteImplementorB)", new Abstraction(new ConcreteImplementorB()).operation());
        assertEquals("RefinedAbstraction(ConcreteImplementorA)",
                new RefinedAbstraction(new ConcreteImplementorA()).operation());
        assertEquals("RefinedAbstraction(ConcreteImplementorB)",
                new RefinedAbstraction(new ConcreteImplementorB()).operation());
    }

    @Test
    void aNewImplementorNeedsNoChangeOnTheAbstractionSide() {
        Implementor custom = () -> "Custom";
        assertEquals("Abstraction(Custom)", new Abstraction(custom).operation());
        assertEquals("RefinedAbstraction(Custom)", new RefinedAbstraction(custom).operation());
    }
}
